cmd/tcplb/server.go

```diff
@@ -0,0 +1,223 @@
+	return IdiotDialer{Logger: logger}, nil
+}
+
+func makeForwarderFromConfig(cfg *Config) (forwarder.Forwarder, error) {
+	// TODO implement something more robust with timeouts
+	return forwarder.MediocreForwarder{}, nil
+}
+
+func coerceIntoAuthenticatedConn(logger slog.Logger, conn net.Conn) (forwarder.AuthenticatedConn, error) {
+	var authenticatedClientConn forwarder.AuthenticatedConn = nil
+	switch cc := conn.(type) {
+	case *tls.Conn:
+		authenticatedClientConn = &authn.AuthenticatedTLSConn{Conn: cc}
+	case *net.TCPConn:
+		logger.Warn(&slog.LogRecord{Msg: "using TCP connection to client, this is insecure"})
+		authenticatedClientConn = &authn.InsecureTCPConn{
+			TCPConn:  cc,
+			ClientID: anonymousTestClientID,
+		}
+	default:
+		return nil, NotImplementedError
+	}
+	return authenticatedClientConn, nil
+}
+
+func serve(logger slog.Logger, cfg *Config) error {
+	// Wire together the forwarder.Server
+
+	reserver, err := makeClientReserverFromConfig(cfg)
+	if err != nil {
+		logger.Error(&slog.LogRecord{Msg: "Client rate-limiter error", Error: err})
+		return err
+	}
+
+	authorizer, err := makeAuthorizerFromConfig(cfg)
+	if err != nil {
+		logger.Error(&slog.LogRecord{Msg: "Authorization configuration error", Error: err})
+		return err
+	}
+
+	dialer, err := makeDialerFromConfig(cfg, logger)
+	if err != nil {
+		logger.Error(&slog.LogRecord{Msg: "Dialer configuration error", Error: err})
+		return err
+	}
+
+	fwder, err := makeForwarderFromConfig(cfg)
+	if err != nil {
+		logger.Error(&slog.LogRecord{Msg: "Forwarder configuration error", Error: err})
+		return err
+	}
+
+	// Compose stack of connection handlers. They are defined
+	// in order from innermost to outermost.
+	forwardingHandler := &forwarder.ForwardingHandler{
+		Logger:    logger,
+		Dialer:    dialer,
+		Forwarder: fwder,
+	}
+	authzHandler := &forwarder.AuthorizedUpstreamsHandler{
+		Logger:     logger,
+		Authorizer: authorizer,
+		Inner:      forwardingHandler,
+	}
+	rateLimitingHandler := &forwarder.RateLimitingHandler{
+		Logger:   logger,
+		Reserver: reserver,
+		Inner:    authzHandler,
+	}
+	recovererHandler := &forwarder.RecovererHandler{
+		Logger: logger,
+		Inner:  rateLimitingHandler,
+	}
+	baseHandler := &forwarder.ConnCloserHandler{
+		Inner: recovererHandler,
+	}
+
+	listener, err := net.Listen(cfg.ListenNetwork, cfg.ListenAddress)
+	if err != nil {
+		msg := fmt.Sprintf("Listen error with network: %s address: %s", cfg.ListenNetwork, cfg.ListenAddress)
+		logger.Error(&slog.LogRecord{Msg: msg, Error: err})
+		return err
+	}
+	defer func() {
+		_ = listener.Close()
+	}()
+
+	// TODO graceful shutdown upon receiving interrupt
+	// - stop accepting new connections
+	// - wait for currently forwarded connections to terminate (hard cut off after timeout?)
+	// - stop healthcheck probes of upstreams (if applicable)
+
+	logger.Info(&slog.LogRecord{Msg: fmt.Sprintf("listening on network: %s address: %s", cfg.ListenNetwork, cfg.ListenAddress)})
+	for {
```

Comment:
I think this should be part of a library package, not package main.

Reply:
good idea. i've moved everything in the for block into the library & simplified the logic a little.

caller is still responsible for supplying a listener & handler.